src/Middleware/WebSockets/src/HandshakeHelpers.cs

```diff
@@ -248,17 +246,69 @@ static bool ParseWindowBits(ReadOnlySpan<char> value, string propertyName, out i
 
             if (!parsedOptions.ServerContextTakeover)
             {
-                builder.Append("; ").Append(WebSocketDeflateConstants.ServerNoContextTakeover);
+                // 2 = '; '
+                responseLength += 2 + WebSocketDeflateConstants.ServerNoContextTakeover.Length;
             }
 
             if (hasServerMaxWindowBits || parsedOptions.ServerMaxWindowBits != 15)
             {
-                builder.Append("; ")
-                    .Append(WebSocketDeflateConstants.ServerMaxWindowBits).Append('=')
-                    .Append(parsedOptions.ServerMaxWindowBits.ToString(CultureInfo.InvariantCulture));
+                // 2 = '; ', 1 = '='
+                responseLength += 2 + WebSocketDeflateConstants.ServerMaxWindowBits.Length + 1 +
+                    ((parsedOptions.ServerMaxWindowBits > 9) ? 2 : 1);
             }
 
-            response = builder.ToString();
+            response = string.Create(responseLength, (parsedOptions, hasClientMaxWindowBits, hasServerMaxWindowBits, hasClientNoContext),
```

Comment:
Another option would be to use https://github.com/dotnet/runtime/blob/5525d0bbb57d5550d71a0a15158f992b0e694000/src/libraries/Common/src/System/Text/ValueStringBuilder.cs

Reply:
Done, looks a bit nicer 😃